### PR TITLE
MGMT-4849 Watch the InfraEnv resource for BMH reconcile

### PR DESCRIPTION
**Pull request by flaper87**
This commit adds a watch on the InfraEnv resource to make sure that we schedule a reconcile when
the ISOUrl has been set. This will protect us from missing this event due to a race or missing
reconcile

Signed-off-by: Flavio Percoco <flavio@redhat.com>

### CODE REVIEW

**Comment by flaper87** [comment]
This will have to be rebased on top of https://github.com/openshift/assisted-service/pull/1584 because of the constant var name change

File: internal/controller/controllers/bmh_agent_controller.go [hunk 1]
```diff
@@ -369,19 +369,16 @@ func (r *BMACReconciler) reconcileBMH(ctx context.Context, bmh *bmh_v1alpha1.Bar
...
 			if err := r.Get(ctx, types.NamespacedName{Name: value, Namespace: bmh.Namespace}, infraEnv); err != nil {
 				return reconcileError{client.IgnoreNotFound(err)}
 			}
 
 			if infraEnv.Status.ISODownloadURL == "" {
-				// the image has not been created yet, try later.
+				// the image has not been created yetlater.
 				r.Log.Infof("Image URL for InfraEnv (%s/%s) not available yet. Retrying reconcile for BareMetalHost  %s/%s",
 					infraEnv.Namespace, infraEnv.Name, bmh.Namespace, bmh.Name)
 				return reconcileRequeue{time.Minute}
```
Comment (rwsu) [inline internal/controller/controllers/bmh_agent_controller.go:384]: Now that there is a watch on InfraEnv, should this still requeue a reconcile?

Comment (flaper87) [reply internal/controller/controllers/bmh_agent_controller.go:441]: Good catch, it shouldn't!

File: internal/controller/controllers/bmh_agent_controller.go [hunk 2]
```diff
@@ -851,7 +851,32 @@ func (r *BMACReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+			return []reconcile.Request{}
+		}
+
+		// Don't queue any reconcile if the InfraEnv
+		// doesn't have the ISODownloadURL set yet.
+		if infraEnv.Status.ISODownloadURL == "" {
+			return []reconcile.Request{}
+		}
+
+		bmh, err := r.findBMHByInfraEnv(ctx, infraEnv)
```
Comment (mhrivnak) [inline internal/controller/controllers/bmh_agent_controller.go:879]: Don't we want this to return a slice of all the BMHs that match the InfraEnv, then transform those into `[]reconcile.Request` ?

File: internal/controller/controllers/bmh_agent_controller.go [hunk 3]
```diff
@@ -438,10 +417,10 @@ func (r *BMACReconciler) reconcileBMH(ctx context.Context, bmh *bmh_v1alpha1.Bar
 			}
 
 			if infraEnv.Status.ISODownloadURL == "" {
-				// the image has not been created yet, try later.
-				r.Log.Infof("Image URL for InfraEnv (%s/%s) not available yet. Retrying reconcile for BareMetalHost  %s/%s",
+				// the image has not been created yetlater.
```
Comment (mhrivnak) [inline internal/controller/controllers/bmh_agent_controller.go:420]: Was this change intentional?

**Review by mhrivnak — CHANGES_REQUESTED** [review]

File: internal/controller/controllers/bmh_agent_controller.go (hunk 3 above)
Comment (flaper87) [reply internal/controller/controllers/bmh_agent_controller.go:420]: it was not, merge conflict gone wrong! :D

File: internal/controller/controllers/bmh_agent_controller.go (hunk 2 above)
Comment (flaper87) [reply internal/controller/controllers/bmh_agent_controller.go:879]: And this, my kids, is the reason why unit-tests are worth almost nothing and everything should have proper functional/integration testing! 

giphy

Great catch!

Comment (flaper87) [reply internal/controller/controllers/bmh_agent_controller.go:879]: oh, hold on! it's actually there L858 but it's collapse because GH diffs are....

Comment (mhrivnak) [reply internal/controller/controllers/bmh_agent_controller.go:879]: This code is finding the first BMH that matches and immediately returning a request just for it. Instead it should continue iterating BMHs, find all of them that match the InfraEnv, and return requests for all of them.

Comment (mhrivnak) [inline internal/controller/controllers/bmh_agent_controller.go:879]: This code is finding the first BMH that matches and immediately returning a request just for it. Instead it should continue iterating BMHs, find all of them that match the InfraEnv, and return requests for all of them.

(trying a new review because just replying did not un-resolve the thread)

**Review by mhrivnak — CHANGES_REQUESTED** [review]

File: internal/controller/controllers/bmh_agent_controller.go (hunk 2 above)
Comment (flaper87) [reply internal/controller/controllers/bmh_agent_controller.go:879]: Oooooh, I see what you mean. I was confused because you commented here, instead of commenting in the `findBMHByInfraEnv` function which, as you correctly suggest, should be returning the list of BMHs associated with a single `InfraEnv`.

Let me fix this, thanks for catching it and explaining it again :pray:

File: internal/controller/controllers/bmh_agent_controller.go [hunk 4]
```diff
@@ -717,6 +696,29 @@ func (r *BMACReconciler) findBMH(ctx context.Context, agent *aiv1beta1.Agent) (*
...
+	if err != nil {
+		return nil, err
+	}
+
+	bmhs := []*bmh_v1alpha1.BareMetalHost{}
+
+	for _, bmh := range bmhList.Items {
+		if val, ok := bmh.ObjectMeta.Labels[BMH_INFRA_ENV_LABEL]; ok {
+			if strings.EqualFold(val, infraEnv.Name) {
+				bmhs = append(bmhs, &bmh)
```
Comment (mhrivnak) [inline internal/controller/controllers/bmh_agent_controller.go:715]: I think you've fallen victim to [this gotcha](https://www.evanjones.ca/go-gotcha-loop-variables.html). The fix is to use the index value, and change this to:

```
bmhs = append(bmhs, &bmhList.Items[i])
```

File: internal/controller/controllers/bmh_agent_controller.go [hunk 5]
```diff
@@ -862,8 +864,42 @@ func (r *BMACReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+		if infraEnv.Status.ISODownloadURL == "" {
+			return []reconcile.Request{}
+		}
+
+		bmhs, err := r.findBMHByInfraEnv(ctx, infraEnv)
+		if len(bmhs) == 0 || err != nil {
+			return []reconcile.Request{}
+		}
+
+		requests := []reconcile.Request{}
```
Comment (mhrivnak) [inline internal/controller/controllers/bmh_agent_controller.go:886]: Minor tweak: since you know the length, you can go ahead and `make` this the correct size, allocating the final structure all at once. Then you can just assign values to each location in the slice instead of using `append`.

**Review by mhrivnak — CHANGES_REQUESTED** [review]

File: internal/controller/controllers/bmh_agent_controller.go (hunk 4 above)
Comment (flaper87) [reply internal/controller/controllers/bmh_agent_controller.go:715]: :facepalm:  thanks

**Review by mhrivnak — APPROVED** [review]
